tls/tls_internal.h

```diff
@@ -135,13 +135,22 @@ typedef struct tls_handshake_t {
 	ttls_sig_hash_set_t		hash_algs;
 	int				sni_authmode;
 
+	unsigned char			ecdh_point_format;
 	unsigned char			point_form		: 1,
```

Comment:
This is probably good to debug, but now in production code we have point format here, in handshake descriptor, and in ECDH context - this is confusing and it simply has no sense for production code. It'd be better to make some nicer solution or at least move the field under `ifdef DEBUG`.

Reply:
It wasn't for debugging. This new `ecdh_point_format` field is to avoid (possible) data corruption.

When we are receiving handshake data, we actively use temporary sha256 context which lives where `ecdh_ctx` is located. But also during handshake processing we use `ecdh_ctx` in `tls->hs->ecdh_ctx.point_format = p[0];`. That way we may accidentally change some byte in sha256 context while it's still used.

I've just checked the numbers again, and it looks like my hypothesis about field access was wrong all the time. Since `offsetof(ttls_ecdh_context, point_format)` is `376`, but `sizeof(ttls_sha256_context)` is only `120`, stores to `ecdh_ctx.point_format` will never touch sha256 context.

So, it turned out the PR doesn't fix any bug. But I'd like to keep it, as it helps to reduce anxiety. At the current moment `ecdh_ctx.point_format` field is at the safe place. However if somebody later tries to reorder `ttls_ecdh_context` struct, and moves `point_format` out of the safe zone, a bug will appear. To avoid that, we need to store `point_format` somewhere before we stop using `ecdh_ctx` memory for anything other than `ecdh_ctx`.

Reply:
The data structures reusage is a hack. I had made several tries to reuse different data structures before I came to ECDH descriptor and I was still wrong. I chose ECDH descriptor because it didn't require additional changes crucial for the code design and clarity (of course any memory reusage is doubtful in terms of code clarity, but it's still a common practice, so that's fine).

I propose to look for other data structures which can be used to place the second checksum. On in this case I'd prefer just to make a comment about special requirements to the data structure members placement.

> But I'd like to keep it, as it helps to reduce anxiety.

I love the macros which make the pointers cast nicer and debugging code which helps the type conversions tracking, but I still do not like the double ecdh point format games.

> Oh, and there is a gap in the struct around there, so adding another unsigned char doesn't increase structure size.

That's fine, but it's still very confusing to seem the same ecdh point format in 2 places.

Reply:
> very confusing to seem the same ecdh point format in 2 places.

Removed that field, reverted to the version without separate field. (See 1254).